include/openssl/bio.h

```diff
@@ -156,10 +156,10 @@ extern "C" {
 
 # ifndef OPENSSL_NO_KTLS
 #  define BIO_get_ktls_send(b)         \
-     (BIO_method_type(b) == BIO_TYPE_SOCKET \
+     (BIO_method_type_final(b) == BIO_TYPE_SOCKET \
       && BIO_ctrl(b, BIO_CTRL_GET_KTLS_SEND, 0, NULL))
 #  define BIO_get_ktls_recv(b)         \
-     (BIO_method_type(b) == BIO_TYPE_SOCKET \
+     (BIO_method_type_final(b) == BIO_TYPE_SOCKET \
```

Comment:
I am starting to have second thoughts about this - and actually to question the original change that I made here. What we are attempting to do is second guess what kind of BIO we have and only send the BIO_CTRL_GET_KTLS_SEND/BIO_CTRL_GET_KTLS_RECV message if the BIO is of the correct type to understand it. But really this should work with any type of BIO. If the BIO doesn't understand the message then it should just fail. As it currently stands no-one could write a custom BIO that intercepts these messages and does something else with them (which might be a desirable property).

I recall there was some problem with the boring tests that caused me to make this change. I think the CTRL messages were getting confused with some other ctrl message, by I forget the details. Probably I need to go back and have another look at the original motivation for this change.

Reply:
Sounds good to me. Let me know if you have an alternative approach and you'd like some help testing it with KTLS.

Reply:
> doesn't the BIO just pass that unknown stuff down to the next BIO in the stack?

Filter BIOs do (well they are not obliged to, but all filter BIOs that we have in the source code do this) - but ultimately it will get down to a source/sink BIO which will either understand the message or not. If it doesn't understand it, it should fail.